cli/src/cli.rs

```diff
@@ -0,0 +1,80 @@
+use crate::config::{Config, FromFile, LoggerConfig};
+use clap::{crate_authors, crate_version, Parser};
+use error_stack::ResultExt;
+pub use error_stack::{Context, Report};
+use std::{fmt::Display, path::PathBuf};
+
+lazy_static::lazy_static! {
+    static ref VERSION: String = {
+        format!(
+            concat!("BOB-GUI VERSION: {}\n",
+            "BUILT AT: {}\n",
+            "COMMIT HASH: {}\n",
+            "GIT BRANCH/TAG: {}\n"),
+            crate_version!(),
+            build_time::build_time_utc!(),
+            option_env!("GIT_HASH").unwrap_or("-"),
+            option_env!("GUI_BUILD_BRANCH_TAG").unwrap_or("-"),
+        )
+    };
+}
+
+/// Bob configuration
+#[derive(Debug, Parser, Clone)]
+#[command(author = crate_authors!())]
+#[command(version = VERSION.trim(), about, long_about)]
+#[group(id = "configs", required = true, multiple = false)]
+pub struct Args {
+    /// If set, passes default configuration to the server
+    #[clap(short, long)]
+    default: bool,
+
+    /// Server configuration file
+    #[arg(short, long, value_name = "FILE")]
+    config_file: Option<PathBuf>,
+}
+
+impl TryFrom<Args> for Config {
+    type Error = Report<Error>;
+
+    fn try_from(value: Args) -> Result<Self, Self::Error> {
+        if value.default {
+            Ok(Self::default())
+        } else if let Some(config) = value.config_file {
+            Self::from_file(config).change_context(Error::Config)
+        } else {
+            unreachable!()
```

Comment:
Why is this `unreachable`? If user do not specify any argument, then this branch will be taken. In this case error should be returned

Reply:
Clap manages input variants on `Args` declaration:
```rust
/// Bob configuration
#[derive(Debug, Parser, Clone)]
#[command(author = crate_authors!())]
#[command(version = VERSION.trim(), about, long_about)]
#[group(id = "configs", required = true, multiple = false)] // <-- Here
pub struct Args {
    /// If set, passes default configuration to the server
    #[clap(short, long)]
    default: bool,

    /// Server configuration file
    #[arg(short, long, value_name = "FILE")]
    config_file: Option<PathBuf>,
}
```
With this line:
```rust
#[group(id = "configs", required = true, multiple = false)] 
```
Clap checks what arguments were provided on startup and if there is none or multiple variants, the error will occur before the binary will even run. I failed to trigger this line no matter what I tried, so I think it's fair to assume that this line is really *unreachable*

Reply:
Why `Args` is not a [`ValueEnum`](https://docs.rs/clap/latest/clap/trait.ValueEnum.html)? The reason is `ValueEnumParser` does not provide parser for non-unit variants afaik, so the config file argument couldn't be passed in this case

Reply:
Ok, didn't notice that. Thanks!